src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,227 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/data-model/WrappedStructEncoder.h>
+#include <lib/core/CHIPError.h>
+
+namespace chip {
+namespace app {
+namespace InteractionModel {
+
+/// Aids in writing invoke replies:
+///   - Has access to the raw data encoder for response data content (via `ResponseEncoder`)
+///   - `Complete` handling:
+///      - MUST be called exactly once
+///      - called with an error, makes TLV data written to be discarded (Invokes may only either
+///        succeed or fully fail)
+class InvokeResponder
+{
+public:
+    virtual ~InvokeResponder() = default;
+
+    // Copying not allowed since underlying requirement is that on deletion of this
+    // object, a reply will be sent.
+    InvokeResponder(const InvokeResponder &)             = delete;
+    InvokeResponder & operator=(const InvokeResponder &) = delete;
+
+    /// Flush any pending replies before encoding the current reply.
+    ///
+    /// MAY be called at most once.
```

Comment:
I am having a little trouble understanding this `MAY be called at most once` I can be called once by an instance of AutoCompleteInvokeResponder, but across multiple instance can this be called more than once?

Let's take a single InvokeRequestMessage that contains 3 batch commands. My understanding is that there will be:
* 1 Instance of `InvokeResponder`
* Up to 3 instance of `AutoCompleteInvokeResponder` all of which may need to call `FlushPendingResponses`

Reply:
My understanding was incorrect. There will be 3 instances of `InvokeResponder`